### PR TITLE
Tanium output fix

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23064

## Description
Added formatting functions to handle the mis-shaped lines

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 1]
```diff
@@ -41,13 +38,52 @@
...
+
+        if line_values[-1].endswith('\"'):
+            line_values[-1] = line_values[-1][:-1]
+
+        if len(line_below) == 1:
+            continue
+
+        else:
+            line_below = line_below[1:]
+            line_values.extend(line_below)
```
Comment (ronykoz) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:59]: not sure what's happening here, please add notes

File: Packs/Legacy/Integrations/Tanium/Tanium_test.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import demistomock as demisto
+
+CSV_STRING_1 = "header1,header2,header3\r\ncol1,col2,col3\r\ncol1,col2,col3"
+CSV_STRING_2 = "header1,header2,header3\r\ncol1,\"col2_1\r\ncol2_2\",col3\r\ncol1,col2,col3"
+CSV_STRING_3 = "header1,header2,header3"
```
Comment (ronykoz) [inline Packs/Legacy/Integrations/Tanium/Tanium_test.py:5]: have these in multiline - or in a separate file altogether

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 1 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: made it clearer in next versions

File: Packs/Legacy/Integrations/Tanium/Tanium_test.py (hunk 2 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium_test.py]: done

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 3]
```diff
@@ -41,13 +38,109 @@
...
+    if len(line_values) == 0 or (len(line_values) == 1 and len(line_values[0]) == 0):
+        current_line += 1
+        return [], current_line
+
+    # multi lined cells start with " and end with "
+    while line_values[-1].startswith("\"") and not line_values[-1].endswith("\""):
+        current_cell = line_values[-1]
+        current_cell, rest_of_line, current_line = get_full_cell(lines, current_line, current_cell)
+        line_values[-1] = current_cell
+        line_values.extend(rest_of_line)
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:105]: Does this logic break a multi line cell into several cells?
If so let's move it to an outer function with that / similar name

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 4]
```diff
@@ -403,6 +496,27 @@ def approveSavedAction(handler, action_id, saved_action_id):
...
+
+    Returns:
+        list. A list of dictionaries where multi-lined cells have spaces instead of line drops.
+    """
+    context = []
+    for element in res:
+        element_context = {}
+
+        for key in element.keys():
+            element_context[key] = element[key].replace("\n", " ")
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:513]: use dict comprehension

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 5]
```diff
@@ -41,13 +38,109 @@
...
+    """
+    lines = text_content.splitlines()
     if len(lines) < 2:
         return []
-    else:
-        headers = lines[0].split(',')
-        return [dict(zip(headers, line.split(','))) for line in lines[1:]]
+
+    headers = lines[0].split(',')
+    total_column_num = len(headers)
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:124]: What if there are  columns with no headers?

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 6]
```diff
@@ -41,13 +38,109 @@
...
+    lines = lines[1:]
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:125]:
```suggestion
    lines = lines[1:] # The first line is the headers
```

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 7]
```diff
@@ -41,13 +38,109 @@
...
+    current_line = 0
+    while current_line < total_lines_num:
+        line_dict = {}
+        line_values, current_line = gather_line(lines, current_line)
+
+        if len(line_values) == 0:
+            continue
+
+        for index in range(total_column_num):
+            line_dict[headers[index]] = line_values[index]
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:137]: why not `for header in headers`?

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 8]
```diff
@@ -41,13 +38,109 @@
...
-        headers = lines[0].split(',')
-        return [dict(zip(headers, line.split(','))) for line in lines[1:]]
+
+    headers = lines[0].split(',')
+    total_column_num = len(headers)
+    lines = lines[1:]
+    total_lines_num = len(lines)
+    result = []
+    current_line = 0
+    while current_line < total_lines_num:
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:129]: Why not use a for loop here?

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 9]
```diff
@@ -41,13 +38,109 @@
...
+        line_dict = {}
+        line_values, current_line = gather_line(lines, current_line)
+
+        if len(line_values) == 0:
+            continue
```
Comment (guykeller) [inline Packs/Legacy/Integrations/Tanium/Tanium.py:134]: if `len(line_values) == 0` you don't advance current_line, won't you get stuck in an infinite loop?

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 10]
```diff
@@ -7,11 +6,9 @@
...
 from cStringIO import StringIO
 
-if not demisto.params()['proxy']:
-    del os.environ['HTTP_PROXY']
-    del os.environ['HTTPS_PROXY']
-    del os.environ['http_proxy']
-    del os.environ['https_proxy']
+handle_proxy(demisto.params().get('proxy'))
+
+response = ''
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py:11]: yes. this was a global variable earlier when there was no "main" function so this is where is should be located now.

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 11]
```diff
@@ -41,13 +38,109 @@
...
+    Args:
+        cell (str): the value of the multi-lined single cell in the table
+
+    Returns:
+        str. The value of the multi-lined cell without the "
+    """
+    return cell[1:-1]
+
+
+def get_full_cell(lines, current_line, current_cell):
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: done

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 12]
```diff
@@ -41,13 +38,109 @@
...
+
+    Args:
+        lines (list): a list of all lines in the csv
+        current_line (int): the current line being gathered
+        current_cell: the value of the multi-lined single cell in the table
+
+    Returns:
+        Tuple(str, list). The full value of the single multi-lined cell and the rest of the values in it's line if exist
+    """
+    current_line = current_line + 1
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: done

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 13]
```diff
@@ -41,13 +38,109 @@
...
+    line_below = lines[current_line].split(',')
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: no this is the line blow the line where we started. current_line (now current_line_index) should point to the line of code we are currently processing so it should move forward - hence the line above.

File: Packs/Legacy/Integrations/Tanium/Tanium.py [hunk 14]
```diff
@@ -41,13 +38,109 @@
...
+    """gather the values of a single line in the answered csv
+
+    Args:
+        lines (list): a list of all lines in the csv
+        current_line (int): the current line being gathered
+
+    Returns:
+        Tuple(list, int). A list containing all the values of a single line, the number of the last line checked
+    """
+    line_values = lines[current_line].split(',')
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: done

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 3 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: It does not break it to several cells - it is still a single cell 

look at this example: 
"header1,header2,header3\r\n" \
"col1,\"col2_1\r\n" \
 "col2_2\r\n" \
"col2_3\",col3" \

you want col2_1 col2_2 and col2_3 to be in the same cell. the rest_of_line is used to handle the col3

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 5 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py:124]: doesn't work that way in the Tanium response

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 6 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: done

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 8 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py]: cause you might jump several lines in a single run of the while loop.

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 9 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py:134]: current_line will be advanced in gather_line specifically in line 97 which handles this case exactly. 
I can move it here if you think this is better.

File: Packs/Legacy/Integrations/Tanium/Tanium.py (hunk 7 above)
Comment (roysagi) [reply Packs/Legacy/Integrations/Tanium/Tanium.py:137]: not sure this looks better:
for header in headers:
     line_dict[header] = line_values[headers.index(header)]

**Comment by roysagi** [comment]
PR got corrupted - reopening in: https://github.com/demisto/content/pull/6720

**Pull request closed without merging**